### PR TITLE
refactor(major): file organization, build process

### PR DESCRIPTION
**Pull request by sarbbottam**
I guess, we both have been working in parallel for the same goal, ~~and thus there are merge conflicts~~.

~~Let's not worry about that right now.~~ I would request you to review the code. Before you start reviewing, allow me to brief you about the changes.
## Overview of changes.
- Refactor src/inject.js, please refer src/index.js
- use webpack instead of gulp
  - webpack will take care of the combining multiple js files as well as the css
  - no need to copy the css files
- chrome/src/inject.js, chrome/src/inject.js.map, firefox/data/inject.js & firefox/data/inject.js.map will be added to .gitignore, before merging the PR.

### CODE REVIEW

File: chrome/manifest.json [hunk 1]
```diff
@@ -24,12 +24,9 @@
       ],
       "all_frames": false,
       "run_at": "document_end",
-      "css": [
-        "src/inject/inject.css"
-      ],
       "js": [
-        "src/inject/inject.js"
+        "src/inject.js"
```
Comment (sarbbottam) [inline chrome/manifest.json:28]: `inject` directory is not required

File: chrome/manifest.json [hunk 2]
```diff
@@ -24,12 +24,9 @@
       ],
       "all_frames": false,
       "run_at": "document_end",
-      "css": [
-        "src/inject/inject.css"
```
Comment (sarbbottam) [inline chrome/manifest.json:28]: `inject.css` is part of `inject.js`, see `src/index.js`

File: chrome/src/inject/inject.css [hunk 3]
```diff
@@ -1,30 +0,0 @@
-.contrib-column {
```
Comment (sarbbottam) [inline chrome/src/inject/inject.css:1]: `inject.css` is part of `inject.js`, see `src/index.js`

File: chrome/src/inject/inject.js [hunk 4]
```diff
@@ -1,159 +0,0 @@
-"use strict";
```
Comment (sarbbottam) [inline chrome/src/inject/inject.js:1]: `inject` directory is not required, see `chrome/src/inject.js`

eventually `chrome/src/inject.js` will be `.gitignore`d

File: firefox/data/inject.css [hunk 5]
```diff
@@ -1,30 +0,0 @@
-.contrib-column {
```
Comment (sarbbottam) [inline firefox/data/inject.css:1]: `inject.css` is part of `inject.js`, see `data/index.js`

File: firefox/index.js [hunk 6]
```diff
@@ -1,9 +1,8 @@
 var self = require("sdk/self");
 var pageMod = require("sdk/page-mod");
- 
+
 pageMod.PageMod({
     include: "https://github.com/*",
     contentScriptFile: self.data.url("inject.js"),
-    contentScriptWhen: "ready",
-    contentStyleFile: self.data.url("inject.css")
-});
```
Comment (sarbbottam) [inline firefox/index.js:9]: `inject.css` is part of `inject.js`, see `data/index.js`

File: package.json [hunk 7]
```diff
@@ -4,13 +4,10 @@
   "description": "Get back longest and current streak on Github prifiles",
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "lint": "eslint .; exit 0",
-    "install": "npm i -g babel-cli watch jpm",
```
Comment (sarbbottam) [inline package.json:8]: `jpm` is part of `devDependencies`

File: package.json [hunk 8]
```diff
@@ -4,13 +4,10 @@
...
-    "clean": "find build -regextype posix-egrep -regex \".*\\.(zip|xpi|nex)$\" -type f -delete && rmdir build",
-    "build:js": "babel ./src/inject.js --presets es2015 | tee ./firefox/data/inject.js > ./chrome/src/inject/inject.js ",
-    "build:css": "echo ./firefox/data/ ./chrome/src/inject/ | xargs -n 1 cp ./src/inject.css",
-    "build": "npm run build:js && npm run build:css",
-    "build:watch": "watch \"npm run build\" ./src/",
+    "lint": "eslint .",
```
Comment (sarbbottam) [inline package.json:7]: no `exit 0`; must fail if there are lint errors

File: package.json [hunk 9]
```diff
@@ -4,13 +4,10 @@
   "description": "Get back longest and current streak on Github prifiles",
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "lint": "eslint .; exit 0",
-    "install": "npm i -g babel-cli watch jpm",
-    "clean": "find build -regextype posix-egrep -regex \".*\\.(zip|xpi|nex)$\" -type f -delete && rmdir build",
```
Comment (sarbbottam) [inline package.json:9]: see `prebuild`

File: src/index.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
+require('./index.css')
```
Comment (sarbbottam) [inline src/index.js:1]: `index.js` is the main file, consider it as entry point.

File: src/index.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+ * to inject github streak container
+ */
+const observer = new MutationObserver((mutations) => {
+    mutations.forEach((mutation) => {
+        if (mutation.oldValue && mutation.oldValue.indexOf("is-loading") > -1 ) {
+            // invoke inject function
+            inject()
+        }
+    })
+})
```
Comment (sarbbottam) [inline src/index.js:24]: mutation observer logic is now simple, it aways observes `js-pjax-loader-bar` and invokes `inject` if desired.

File: src/inject.js [hunk 12]
```diff
@@ -121,50 +119,4 @@ function inject() {
...
-                        inject()
-                        attachClickInjecter()
-                        observer.disconnect()
-                    }
-                })
-            })
-            observer.observe(pixelBar, config)
-        }, false)
-    }))
-}
```
Comment (sarbbottam) [inline src/inject.js:141]: see `src/index.js`, this logic has been simplified.

File: src/inject.js [hunk 13]
```diff
@@ -121,50 +119,4 @@ function inject() {
...
-        "1e6823": "0363F0"
-    }
-
-    if(color) {
-        Object.keys(colorsHash).forEach((key) => {
-            updateStyle(`rect[fill='#${key}']`, "fill", `#${colorsHash[key]}`)
-            updateStyle(`li[style='background-color: #${key}']`, "backgroundColor", `#${colorsHash[key]}`)
-        })
-    }
-}
```
Comment (sarbbottam) [inline src/inject.js:165]: see `src/change-color.js`

**Comment by Naramsim** [comment]
Well, these are a lot of big and great changes! :smile: 
Thanks for your effort @sarbbottam. Tomorrow I will take a deeper look.

**Comment by sarbbottam** [comment]
> Well, these are a lot of big and great changes!

You bet!! 😄 

Don't hurry, take your time.

**Comment by Naramsim** [comment]
Never heard about Webpack

**Comment by sarbbottam** [comment]
To be honest I prefer a PR not more than 4 files changed or less that 200 LOC. But could not help in this one, they are so tightly coupled.

**Comment by sarbbottam** [comment]
> Never heard about Webpack

It's great, and webpack 2.x is going to be awesome.

**Comment by sarbbottam** [comment]
If you are still there, what is bithound for?

**Comment by sarbbottam** [comment]
I would recommend to [clone the PR](https://github.com/sarbbottam/conf-files/blob/master/dot-commons/functions#L23-L35), as you review, I find it helpful when a PR involves, lot of changes.

I would suggest the following order of files for review:

Simple
- .eslintignore
- .gitignore

Deleted as they will be generated
- chrome/src/inject/inject.css
- chrome/src/inject/inject.js
- firefox/data/inject.css
- firefox/data/inject.js

Refactored
- src/change-color.js
- src/index.css
- src/index.js
- src/inject.js

Generate the inject.js
- webpack.config.js

Updated w.r.t. file and location
- chrome/manifest.json
- firefox/index.js

Dependencies and scripts
- package.json

**Comment by Naramsim** [comment]
Ok, I got it all :smile: 
But there are two things that do not convince me. I will comment the code

File: package.json [hunk 14]
```diff
@@ -4,13 +4,10 @@
...
-    "build:js": "babel ./src/inject.js --presets es2015 | tee ./firefox/data/inject.js > ./chrome/src/inject/inject.js ",
-    "build:css": "echo ./firefox/data/ ./chrome/src/inject/ | xargs -n 1 cp ./src/inject.css",
-    "build": "npm run build:js && npm run build:css",
-    "build:watch": "watch \"npm run build\" ./src/",
+    "lint": "eslint .",
+    "prebuild": "npm run lint && rimraf chrome/src/ firefox/data/ build",
+    "build": "webpack -p",
```
Comment (Naramsim) [inline package.json]: Here Webpack compresses everything. When submitting a file to Opera and Mozilla store they must be able to read it or you can give them the instructions to get the same result with the compressed code.
I realized that Opera does not even try to build it and do not approves your extension. (I tried with vs1.5 and no response for a week)

File: src/index.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+/**
+ * observe pjaxLoaderBar for changes in attributes
+ */
+const pjaxLoaderBar = document.getElementById("js-pjax-loader-bar")
+const config = { attributes: true, attributeOldValue: true }
+observer.observe(pjaxLoaderBar, config)
+
+// invoke changeColor function
+changeColor()
```
Comment (Naramsim) [inline src/index.js]: `changeColor()` must be invoked with `inject()`. If you don't when you click on repositories tab and back to overview, the squares are repainted in green colors

File: package.json (hunk 14 above)
Comment (sarbbottam) [inline package.json]: That's easy, I have to remove `-p` production flag 😄

File: src/index.js (hunk 15 above)
Comment (sarbbottam) [inline src/index.js]: Ahh, make sense, so we can invoke `changeColor()` whenever `inject()` is being invoked.

**Comment by sarbbottam** [comment]
@Naramsim could you take another look?

**Pull request merged**